libgraphics/src/coretext-skia.cpp

```diff
@@ -0,0 +1,418 @@
+    CFRelease(t_line);
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+MCGFloat __MCGContextMeasurePlatformText(MCGContextRef self, const unichar_t *p_text, uindex_t p_length, const MCGFont &p_font, const MCGAffineTransform &p_transform)
+{
+    // The transform that drawing will be using
+    MCGAffineTransform t_transform;
+    if (self != nil)
+        t_transform = MCGContextGetDeviceTransform(self);
+    else
+        t_transform = MCGAffineTransformMakeIdentity();
+    
+    // Create a fully laid-out line of text
+    CTLineRef t_line = MCGCreateCTLineFromText(p_text, p_length, p_font);
+    
+    // Left and right bounds of the line
+    MCGFloat t_left = 0;
+    MCGFloat t_right = 0;
+    
+    // Iterate over the runs of text within the line
+    MCGForEachRunInLine(t_line,
+    [=, &t_left, &t_right](CTRunRef p_run)
+    {
+        // Get the attribute dictionary for this run
+        CFDictionaryRef t_attributes = CTRunGetAttributes(p_run);
+        
+        // Get the CTFontRef used for drawing this run
+        CTFontRef t_font = static_cast<CTFontRef>(CFDictionaryGetValue(t_attributes, kCTFontAttributeName));
+        
+        // Calculate the origin of this run
+        CGAffineTransform t_text_transform = CTRunGetTextMatrix(p_run);
+        MCGFloat t_x = t_text_transform.tx;
+        
+        // Create a Skia paint object wrapping the font
+        SkPaint t_paint = MCGSkPaintFromCTFontRef(t_font, t_transform);
+        
+        // Force anti-aliasing on so that we get nicely-rendered text
+        t_paint.setAntiAlias(true);
+        
+        // We are going to draw text by glyph ID rather than by codeunit
+        t_paint.setTextEncoding(SkPaint::kGlyphID_TextEncoding);
+        
+        // Get the glyphs belonging to this run
+        CFIndex t_glyph_count = CTRunGetGlyphCount(p_run);
+        MCGMaybeOwnedArray<CGGlyph> t_glyphs(t_glyph_count);
+        if (!t_glyphs.TryBorrow(CTRunGetGlyphsPtr(p_run)))
+        {
+            if (t_glyphs.Allocate())
+                CTRunGetGlyphs(p_run, CFRangeMake(0, t_glyph_count), t_glyphs);
+        }
+        
+        // Measure the run
+        // This line depends on the assumption that CGGlyph and SkGlyphID are the same type
+        MCGFloat t_width = t_paint.measureText(t_glyphs, t_glyph_count * sizeof(CGGlyph));
+        
+        // Check for extension of the left-hand bound of the run
+        t_left = MCMin(t_left, t_x);
+        
+        // Check for extension of the right-hand bound of the run
+        t_right = MCMax(t_right, t_x + t_width);
+    });
+    
+    // Release typeset line
+    CFRelease(t_line);
+    
+    // Return the width of the run
+    return t_right - t_left;
+}
+
+// TODO: this seems to have no callers
+bool MCGContextMeasurePlatformTextImageBounds(MCGContextRef self, const unichar_t *p_text, uindex_t p_length, const MCGFont &p_font, const MCGAffineTransform &p_transform, MCGRectangle &r_bounds)
```

Comment:
this is called (eventually) by the LCB canvas module in the implementation of "the image bounds of ..."